File: agent/sdk/python/layerx_sdk/launchpad.py
"""Calldata builders, wallet send helpers and event decoders for the Launchpad precompile."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Literal

from .exchange import (
    AbiType,
    AbiValue,
    DecodedEvent,
    EventInput,
    EventSpec,
    PrecompileCall,
    WalletRequest,
    decode_event_from,
    encode_abi_call,
    send_precompile_call,
)

LAUNCHPAD_PRECOMPILE = "0x0000000000000000000000000000000000001017"

LaunchpadTokenWrite = Literal["claimAirdrop", "executeAirdrop", "executeBurn", "executeLpRewards", "pause", "unpause"]


def _event(name: str, *inputs: tuple[str, AbiType, bool]) -> EventSpec:
    return EventSpec(name, LAUNCHPAD_PRECOMPILE, tuple(EventInput(*item) for item in inputs))


LAUNCHPAD_EVENTS: tuple[EventSpec, ...] = (
    _event("AirdropClaimed", ("token", "address", True), ("holder", "address", True), ("amount", "uint256", False), ("epoch", "uint256", False)),
    _event("AirdropExecuted", ("token", "address", True), ("amount", "uint256", False), ("epoch", "uint256", False)),
    _event(
        "FeeRecorded",
        ("token", "address", True),
        ("feeAmount", "uint256", False),
        ("protocolCut", "uint256", False),
        ("poolCut", "uint256", False),
    ),
    _event("FeeStrategyChanged", ("token", "address", True), ("oldStrategy", "uint8", False), ("newStrategy", "uint8", False)),
    _event("FeesBurned", ("token", "address", True), ("amount", "uint256", False)),
    _event("FeesClaimed", ("token", "address", True), ("recipient", "address", True), ("amount", "uint256", False)),
    _event("LpRewardsExecuted", ("token", "address", True), ("amount", "uint256", False)),
    _event(
        "MarketCreated",
        ("token", "address", True),
        ("creator", "address", True),
        ("denom", "string", False),
        ("name", "string", False),
        ("symbol", "string", False),
        ("feeStrategy", "uint8", False),
    ),
    _event("PauseToggled", ("token", "address", True), ("paused", "bool", False)),
    _event(
        "Swap",
        ("token", "address", True),
        ("trader", "address", True),
        ("recipient", "address", True),
        ("isBuy", "bool", False),
        ("amountIn", "uint256", False),
        ("amountOut", "uint256", False),
        ("feeAmount", "uint256", False),
        ("price", "uint256", False),
    ),
)


def decode_launchpad_event(address: str, topics: Sequence[str], data: str) -> DecodedEvent:
    return decode_event_from(LAUNCHPAD_EVENTS, address, topics, data)


def _call(name: str, types: Sequence[AbiType], values: Sequence[AbiValue]) -> PrecompileCall:
    return PrecompileCall(LAUNCHPAD_PRECOMPILE, encode_abi_call(name, types, values))


_SWAP: list[AbiType] = ["address", "uint256", "uint256", "address", "uint256"]


def launchpad_buy_call(token: str, quote_in: int, min_out: int, recipient: str, deadline: int) -> PrecompileCall:
    """``buy(address,uint256,uint256,address,uint256)``."""
    return _call("buy", _SWAP, [token, quote_in, min_out, recipient, deadline])


def launchpad_sell_call(token: str, amount_in: int, min_out: int, recipient: str, deadline: int) -> PrecompileCall:
    """``sell(address,uint256,uint256,address,uint256)``."""
    return _call("sell", _SWAP, [token, amount_in, min_out, recipient, deadline])


def launchpad_create_market_call(name: str, symbol: str, fee_strategy: int) -> PrecompileCall:
    """``createMarket(string,string,uint8)``."""
    return _call("createMarket", ["string", "string", "uint8"], [name, symbol, fee_strategy])


def launchpad_set_fee_strategy_call(token: str, fee_strategy: int) -> PrecompileCall:
    """``setFeeStrategy(address,uint8)``."""
    return _call("setFeeStrategy", ["address", "uint8"], [token, fee_strategy])


def launchpad_claim_fees_call(token: str, recipient: str) -> PrecompileCall:
    """``claimFees(address,address)``."""
    return _call("claimFees", ["address", "address"], [token, recipient])


def launchpad_token_call(write: LaunchpadTokenWrite, token: str) -> PrecompileCall:
    """``claimAirdrop``, ``executeAirdrop``, ``executeBurn``, ``executeLpRewards``, ``pause`` or ``unpause``."""
    return _call(write, ["address"], [token])


def send_launchpad_buy(request: WalletRequest, sender: str, token: str, quote_in: int, min_out: int, recipient: str, deadline: int) -> str:
    return send_precompile_call(request, sender, launchpad_buy_call(token, quote_in, min_out, recipient, deadline))


def send_launchpad_sell(request: WalletRequest, sender: str, token: str, amount_in: int, min_out: int, recipient: str, deadline: int) -> str:
    return send_precompile_call(request, sender, launchpad_sell_call(token, amount_in, min_out, recipient, deadline))


def send_launchpad_create_market(request: WalletRequest, sender: str, name: str, symbol: str, fee_strategy: int) -> str:
    return send_precompile_call(request, sender, launchpad_create_market_call(name, symbol, fee_strategy))


def send_launchpad_set_fee_strategy(request: WalletRequest, sender: str, token: str, fee_strategy: int) -> str:
    return send_precompile_call(request, sender, launchpad_set_fee_strategy_call(token, fee_strategy))


def send_launchpad_claim_fees(request: WalletRequest, sender: str, token: str, recipient: str) -> str:
    return send_precompile_call(request, sender, launchpad_claim_fees_call(token, recipient))


def send_launchpad_token_write(request: WalletRequest, sender: str, write: LaunchpadTokenWrite, token: str) -> str:
    return send_precompile_call(request, sender, launchpad_token_call(write, token))

File: agent/sdk/typescript/src/launchpad.ts
/** Calldata builders, wallet send helpers and event decoders for the Launchpad precompile. */

import type { Eip1193Requester } from "./account-derivation.js";
import {
  decodeEventFrom,
  encodeAbiCall,
  sendPrecompileCall,
  type DecodedPrecompileEvent,
  type PrecompileAbiType,
  type PrecompileAbiValue,
  type PrecompileCall,
  type PrecompileEventInput,
  type PrecompileEventSpec,
  type PrecompileLog,
} from "./exchange.js";

export const LAUNCHPAD_PRECOMPILE = "0x0000000000000000000000000000000000001017";

function event(name: string, inputs: readonly PrecompileEventInput[]): PrecompileEventSpec {
  return { name, precompile: LAUNCHPAD_PRECOMPILE, inputs };
}

function field(name: string, type: PrecompileAbiType, indexed = false): PrecompileEventInput {
  return { name, type, indexed };
}

export const LAUNCHPAD_EVENTS: readonly PrecompileEventSpec[] = [
  event("AirdropClaimed", [field("token", "address", true), field("holder", "address", true), field("amount", "uint256"), field("epoch", "uint256")]),
  event("AirdropExecuted", [field("token", "address", true), field("amount", "uint256"), field("epoch", "uint256")]),
  event("FeeRecorded", [field("token", "address", true), field("feeAmount", "uint256"), field("protocolCut", "uint256"), field("poolCut", "uint256")]),
  event("FeeStrategyChanged", [field("token", "address", true), field("oldStrategy", "uint8"), field("newStrategy", "uint8")]),
  event("FeesBurned", [field("token", "address", true), field("amount", "uint256")]),
  event("FeesClaimed", [field("token", "address", true), field("recipient", "address", true), field("amount", "uint256")]),
  event("LpRewardsExecuted", [field("token", "address", true), field("amount", "uint256")]),
  event("MarketCreated", [
    field("token", "address", true),
    field("creator", "address", true),
    field("denom", "string"),
    field("name", "string"),
    field("symbol", "string"),
    field("feeStrategy", "uint8"),
  ]),
  event("PauseToggled", [field("token", "address", true), field("paused", "bool")]),
  event("Swap", [
    field("token", "address", true),
    field("trader", "address", true),
    field("recipient", "address", true),
    field("isBuy", "bool"),
    field("amountIn", "uint256"),
    field("amountOut", "uint256"),
    field("feeAmount", "uint256"),
    field("price", "uint256"),
  ]),
];

/** Decodes a Launchpad precompile log. */
export function decodeLaunchpadEvent(log: PrecompileLog): DecodedPrecompileEvent {
  return decodeEventFrom(LAUNCHPAD_EVENTS, log);
}

function launchpadCall(name: string, types: readonly PrecompileAbiType[], values: readonly PrecompileAbiValue[]): PrecompileCall {
  return { to: LAUNCHPAD_PRECOMPILE, data: encodeAbiCall(name, types, values), value: 0n };
}

export interface LaunchpadSwapOrder {
  readonly token: string;
  readonly amountIn: bigint;
  readonly minOut: bigint;
  readonly recipient: string;
  readonly deadline: bigint;
}

const SWAP_TYPES: readonly PrecompileAbiType[] = ["address", "uint256", "uint256", "address", "uint256"];

function swapValues(order: LaunchpadSwapOrder): readonly PrecompileAbiValue[] {
  return [order.token, order.amountIn, order.minOut, order.recipient, order.deadline];
}

/** `buy(address,uint256,uint256,address,uint256)`; `amountIn` is the quote paid in. */
export function launchpadBuyCall(order: LaunchpadSwapOrder): PrecompileCall {
  return launchpadCall("buy", SWAP_TYPES, swapValues(order));
}

/** `sell(address,uint256,uint256,address,uint256)`. */
export function launchpadSellCall(order: LaunchpadSwapOrder): PrecompileCall {
  return launchpadCall("sell", SWAP_TYPES, swapValues(order));
}

/** `createMarket(string,string,uint8)`. */
export function launchpadCreateMarketCall(name: string, symbol: string, feeStrategy: number): PrecompileCall {
  return launchpadCall("createMarket", ["string", "string", "uint8"], [name, symbol, BigInt(feeStrategy)]);
}

/** `setFeeStrategy(address,uint8)`. */
export function launchpadSetFeeStrategyCall(token: string, feeStrategy: number): PrecompileCall {
  return launchpadCall("setFeeStrategy", ["address", "uint8"], [token, BigInt(feeStrategy)]);
}

/** `claimFees(address,address)`. */
export function launchpadClaimFeesCall(token: string, recipient: string): PrecompileCall {
  return launchpadCall("claimFees", ["address", "address"], [token, recipient]);
}

export type LaunchpadTokenWrite = "claimAirdrop" | "executeAirdrop" | "executeBurn" | "executeLpRewards" | "pause" | "unpause";

/** `claimAirdrop`, `executeAirdrop`, `executeBurn`, `executeLpRewards`, `pause` or `unpause` on `(address)`. */
export function launchpadTokenCall(write: LaunchpadTokenWrite, token: string): PrecompileCall {
  return launchpadCall(write, ["address"], [token]);
}

export const sendLaunchpadBuy = (wallet: Eip1193Requester, from: string, order: LaunchpadSwapOrder): Promise<string> =>
  sendPrecompileCall(wallet, from, launchpadBuyCall(order));

export const sendLaunchpadSell = (wallet: Eip1193Requester, from: string, order: LaunchpadSwapOrder): Promise<string> =>
  sendPrecompileCall(wallet, from, launchpadSellCall(order));

export const sendLaunchpadCreateMarket = (
  wallet: Eip1193Requester,
  from: string,
  name: string,
  symbol: string,
  feeStrategy: number,
): Promise<string> => sendPrecompileCall(wallet, from, launchpadCreateMarketCall(name, symbol, feeStrategy));

export const sendLaunchpadSetFeeStrategy = (
  wallet: Eip1193Requester,
  from: string,
  token: string,
  feeStrategy: number,
): Promise<string> => sendPrecompileCall(wallet, from, launchpadSetFeeStrategyCall(token, feeStrategy));

export const sendLaunchpadClaimFees = (
  wallet: Eip1193Requester,
  from: string,
  token: string,
  recipient: string,
): Promise<string> => sendPrecompileCall(wallet, from, launchpadClaimFeesCall(token, recipient));

export const sendLaunchpadTokenWrite = (
  wallet: Eip1193Requester,
  from: string,
  write: LaunchpadTokenWrite,
  token: string,
): Promise<string> => sendPrecompileCall(wallet, from, launchpadTokenCall(write, token));
